### ISSUE TITLE
Provide some sort of event system

### ISSUE THREAD

**Issue by torkleyy** [amethyst/amethyst#229]
A more general purpose event system would be nice, probably by changing `handle_events` to receive not just window events.

I'd need this for asset management, where the `State` will be informed when an `AssetSet` has finished.

**Comment by vitvakatu** [comment]
I can work on it. Do you have some ideas?
Or maybe you want to do it yourself?

**Comment by torkleyy** [comment]
@vitvakatu Sure, you can do it, if you want. I hoped to have some discussion about what it should offer. For example, do we want a fixed set of events (enum) or unlimited?

**Comment by vitvakatu** [comment]
I suppose, unlimited (user defined) set of events is better. Another question, could we implement it in pretty way.

**Comment by torkleyy** [comment]
OTOH why would the user emit an event if the only receiver is the active state?

**Comment by Lionex** [comment]
[Looking back at the inspiration,](http://gameprogrammingpatterns.com/event-queue.html) it seems prudent to provide a way to create new event queues and listeners for things like the audio system, or potentially a tutorial system, combat system, etc.

**Comment by vitvakatu** [comment]
Hmm... @torkleyy you're write. Messages (or events, it's the same) are belong to OOP, it's not the Rust way. We should have specs and our state system in mind while planning the event system.

**Comment by OvermindDL1** [comment]
As an example, since I keep popping up here and want amethyst to eventually replace my custom-made C++ ECS engine, I guess I should state how my events worked.  :-)

First the gist, the event system had a couple of parts, but the same basic interface for it all, and it was designed to be both very efficient (some paths could even be inlined by the compiler for static knowledge) as well as easy to use via scripting interfaces.

Internally it had a few parts, the first of which was the static dispatching, these were for performance sensitive parts used by things like collision, this was fast, low level, but could still be interacted through via the generic dynamic interfaces.  The rest was built on top of the static dispatching with a generic path for generic events that were entirely arbitrary and user defined, the interface through this part was the same no matter the event type, which was to send:
```cpp
context.sendEvent(Atom eventName, VariantMap &eventData);
// -- or --
context.sendEvent(NodePtr node, Atom eventName, VariantMap &eventData);
// -- or --
context.sendEvent(EntityID entity, Atom eventName, VariantMap &eventData);
```
Callback functions were always of the form:
```cpp
void callbackFunction(Atom eventName, VariantMap &eventData);
```
And registration would be something like:
```cpp
EventCallbackID id = context.registerEventCallback(Atom eventName, CallbackType callback);
// -- or --
EventCallbackID id = context.registerEventCallback(NodePtr node, Atom eventName, CallbackType callback);
// -- or --
EventCallbackID id = context.registerEventCallback(EntityID entity, Atom eventName, CallbackType callback);
```
The CallbackType was just a normal C++ `std::function<void(Atom, VariantMap&)>` so you can bind variables, make closures, etc... etc...  You can unregister by called `context.unregisterEventCallback(id);`.

The node variants only registered/sent events in that tree node path (as I described how it works in other issues), and the entity version did it only on that specific entity (the event's were just another component for entities with an event ECS system handling them), good to listen to if a game lever was switched or something too damage or so.

The VariantMap was just a red/block-tree map (since they are often very small in key count, <10 or so on average) of the accepted marshallable types to the scripting system and so forth (detailed elsewhere too).

The Atom type I've also described elsewhere, and I happen to have it on [github](https://github.com/OvermindDL1/OverECS/blob/master/StringAtom.hpp) (see the comment at the bottom for usage), but it is just an integer that is internally a packed restricted string, but being an integer it is wonderfully fast, and the scripting side can be a little slower by using strings directly too if they want.

**Comment by torkleyy** [comment]
@OvermindDL1 

I'll try to summarize what you're saying, please correct me if I got it wrong.

* Not just `State`s receive events, but everything can register a callback
* Custom events can be emitted because the events are essentially small stack-allocated strings, having no complications like numbers (which you have to manage somehow, which isn't really convenient), while still being very efficient
* `Entity`s can also receive events through a separate event component

Unfortunately, I don't know what nodes you are talking about, could please clarify that or point me to the relevant discussion?

**Comment by vitvakatu** [comment]
I think, `entity` cannot receive events, because it's basically against ECS. Components cannot store code, it's just pure data. How component can work with events? I have no idea.

**Comment by torkleyy** [comment]
@vitvakatu you are right, it's not intended by design! It's possible, because you can just store a function pointer, but now that you brought that up, it feels like a hack.

**Comment by OvermindDL1** [comment]
> Not just `States` receive events, but everything can register a callback

Correct, it accepted an arbitrary C++ decorated function pointer (which can be substantially larger than just a function pointer in terms of functionality).

> Custom events can be emitted because the events are essentially small stack-allocated strings, having no complications like numbers (which you have to manage somehow, which isn't really convenient), while still being very efficient

No strings actually, an atom is just a 64-bit integer, nothing else, I just used C++11 compile-time metaprogramming to convert strings into the integers, however there are also runtime conversions of the strings to integers and back again, and it throws errors if it is an invalid string for the encoding.  For example there is a global even with the atom of `"UPDATE"_atom64`, which is just a 64-bit integer at runtime, which things can register to for handling per-frame updates, like polling the network system for packets to dispatch.  I always meant to add in a priority system to the events but I never got around to it, instead I just ended up adding things like PREUPDATE and POSTUPDATE and such events...  >.>

> `Entitys` can also receive events through a separate event component

Yep, this was *HIGHLY* useful.

> Unfortunately, I don't know what nodes you are talking about, could please clarify that or point me to the relevant discussion?

The Nodes were in my filesystem like tree in my engine, everything was registered to a point in it so, say, a hardware keyboard was at `/dev/keyboards/<UUID>` and an in-game lever that opens the boss door might be available at both `/currentlevel/ecs/entities/42` and `/currentlevel/ecs/named/boss_lever` and so forth, it is not related to events (although you can set connections on it like events or an Rx system).

> I think, `entity` cannot receive events, because it's basically against ECS. Components cannot store code, it's just pure data. How component can work with events? I have no idea.

My components were always pure data, not even pointers, nothing that cannot be serialized by a basic memcpy, the events for entities were done via a separate system, it was just abstracted so it appeared in the normal ECS, but it was often used for things like registering an DAMAGED event on a specific entity by a component system when a, say, ExplodeOnDeath component was added, so that after the damage system processed the damages applied to this entity then it sent out a DAMAGED event on the entity itself and that other system then reacts to it by also blowing up.  Or one entity could listen to the DAMAGED event on another entity and do something like attack the attacker or so.

**Comment by torkleyy** [comment]
Summary of what was discussed on Gitter so far:

Provide an enum, `Event`:

```rust
pub enum Event {
    Asset(assets::Event),
    System(Something),
    Window(winit::Event),
    User(String, Box<Any>),
}
```
* Events are processed asynchronously, so you can easily do expensive work when receiving an event
* Events are also passed to the renderer which allows creating a mesh from wherever you want

**Comment by torkleyy** [comment]
It turns out this is not as easy as I wrote above. For one, I wrote that they should be processed asynchronously; however, this means we spawn one closure into the thread pool per event. Do we want that? If not, how should events be handled?
Another issue I'm not yet sure about is how many events do we want to have? How should one tell if something should emit an event?

cc @ebkalderon

**Comment by fuchsnj** [comment]
@torkleyy I'm thinking we would want the events to be processed synchronously. I'd imagine most events generally need to be processed the same frame they were generated. If this was async, we would need to provide a way to block rendering the frame until all events that you want to complete that frame are done.
Most events are probably quick to process and won't be a problem. Any events that a user actually wants to process async could be done fairly trivially themself (with a threadpool or something).

If we did want async events, I would look into `futures:stream:Stream` from the `futures` crate. Amethyst could provide the Stream and populate it with events, but it's up to the user how the events are consumed.

**Comment by OvermindDL1** [comment]
Yeah I kept my events entirely synchronous.  As in the second an event was placed then all listeners were called before it even returned.  This was useful for two reasons, first is that sometimes I wanted things to 'return' data from an event (I'd have a `return` field or `canceled` field or so on the event structure), and second was that if I need async then that callback could just do as little possible to push it into something else for later processing.

Really though, do not underestimate the ability to return data from event callbacks on the event structure, it is amazingly useful.

**Comment by torkleyy** [comment]
Thanks for the replies so far, your arguments included some points I already though about, but I have some questions and there are issues which were not addressed so far.

@fuchsnj 

> If this was async, we would need to provide a way to block rendering the frame until all events that you want to complete that frame are done

In which situations is it necessary to have something finish before the next frame?

> If we did want async events, I would look into futures:stream:Stream from the futures crate. Amethyst could provide the Stream and populate it with events, but it's up to the user how the events are consumed.

We sort of decided to avoid futures, especially in this situation. It isn't very practical to have a stream of events, especially because you want an event to notify you, not ask if something happened (then we wouldn't need events in the first place).

@OvermindDL1 

> sometimes I wanted things to 'return' data from an event (I'd have a return field or canceled field or so on the event structure)

I already considered this as a pro of synchronized events, but I wasn't aware of the actual usefulness. Thanks for sharing this experience!

---

## Issues which were not yet addressed

### Scripting

I promised to open an issue for this, but I didn't yet find the time to do so. The rough idea for a scripting language is that it works event based, so a script registers itself for some event and it will be called then. Depending on how this is going to work out, we definitely want to provide a way to let the script execute something asynchronously.

### What is considered an event?

First of all, an "event" is not precisely defined.

In my opinion, events should only be used to notify about something, like a key press, an asset that was loaded, a player just died, ...
It should *not* be used however to explicitly start an action, like "load an asset".

### How to register an event listener?

I would not go for callbacks. Instead, we could register an enum `Receiver` which may look like this:

```rust
pub enum Reveicer {
    /// The active state
    State,
    Script(SomeSortOfScriptHandle),
    ...
}
```

A question I'm not yet sure about is how events should relate with states. Should we remove the `State` from `Receiver` and handle it in a different way? I'm convinced however that only the active state should receive events.

**Comment by minecrawler** [comment]
> What is considered an event?

I think, [Bob Nystrom](http://gameprogrammingpatterns.com/event-queue.html) explains it very well:

> An “event” or “notification” describes something that already happened, like “monster died”. You queue it so that other objects can respond to the event

[..]

> A “message” or “request” describes an action that we want to happen in the future, like “play sound”

As for state: I think that it might sometimes make sense to have the state of the point in time when the message was fired. However, on other occasions, you might need the current state. Example:

- You might need the past state when handling keyboard input, as the keys might not be pressed any more
- You might need the current state for AI decisions, as they change all the time, so the most current situation is applicable.

Based on that, it might make sense to send the old state. If the event handler needs the current state, it can query it itself in a frame-blocking operation. Alternatively: Implement two different handlers, one for frame-independent operations, one for frame-dependent operations.

**Comment by torkleyy** [comment]
@minecrawler Yes, that explanation matches pretty much with what I wanted to say.

I disagree that the previous state should receive keyboard input, that would mean that you are in the pause menu (so a state on top of the game state), you press your left mouse button and the underlying state receives the event, so your player fires his weapon, which is not quite what we want.

**Comment by minecrawler** [comment]
@torkleyy The state should also contain the game state, for example "are we in a menu?" "are we in a certain zone in-game?", etc. That stuff, however, is application-specific and we should not care about it in Amethyst at all.

**Comment by torkleyy** [comment]
@minecrawler

> The state should also contain the game state, for example "are we in a menu?"

Why should a state know which state we are in? If Amethyst wouldn't care about states, why do we have a state machine?

The principle of having states is not game specific, but included in many, many games. States should be handled with some assumption, so it is easier for the user.

**Comment by OvermindDL1** [comment]
> I think, Bob Nystrom explains it very well:

What if you wanted to respond to the event by, say, canceling the death of the monster (special ability or so, free life)?

I guess it is more of a 'hook' then, but I use those styles probably more than I really should in my C++ engine, they are just so amazingly useful.

> I disagree that the previous state should receive keyboard input, that would mean that you are in the pause menu (so a state on top of the game state), you press your left mouse button and the underlying state receives the event, so your player fires his weapon, which is not quite what we want.

Remember where I spoke of my systemtree that could register and unregister events?  This is the place where it is easy to bind an input system to something else or vice-versa back or to cancel an event in progress so it does not propagate further.

> ###Scripting
> 
> I promised to open an issue for this, but I didn't yet find the time to do so. The rough idea for a scripting language is that it works event based, so a script registers itself for some event and it will be called then. Depending on how this is going to work out, we definitely want to provide a way to let the script execute something asynchronously.

Also remember that I showed in my event system how an event 'name' was an atom (64-bit integer, but a string in source code) and the payload was a VariantMap?  This was chosen specifically for ease of scripting and modding and it is still blazing fast (as long as the payload is under 10 or so element, but even higher is still faster than a linear lookup anyway).  I'd not change this part of my style either.

**Comment by minecrawler** [comment]
@torkleyy I fully agree. Amethyst needs some kind of state for itself. However it does not need to know about player health. What if there are several players? What if there are none? Amethyst needs to know about the state of resources, not about the state of the game. However, since the game reacts to certain events, it might need the additional game state at the point in the past.

To sum it up, the state contains many different things. Some are useful, others are not. The logic has to decide what things it needs.


> What if you wanted to respond to the event by, say, canceling the death of the monster (special ability or so, free life)?

Which does not mean that the monster did not die. The monster's state will just be modified to be alive (with a certain amount of HP). Then, a revival-sound message can be fired and the AI will fetch the most current state on the next update and control the monster as though nothing happened.

**Comment by torkleyy** [comment]
@minecrawler Could you present an example where you'd need to receive an event with another system than the current one? I mean, of course Amethyst does not need to know the state of the game, but the states in the state machine *are* representing the game state (or are supposed to be).

**Comment by OvermindDL1** [comment]
> @minecrawler Could you present an example where you'd need to receive an event with another system than the current one? I mean, of course Amethyst does not need to know the state of the game, but the states in the state machine are representing the game state (or are supposed to be).

Well a scripting system would be one for sure, it would want to register to and listen to events all over the place depending on what the scripts want to listen to.

**Comment by minecrawler** [comment]
Stuff which happens asynchronous to the whole ECS, for example networking, asset loading,... maybe game logic.

**Comment by torkleyy** [comment]
@OvermindDL1 Regarding the cheap strings you were describing, we could use [`smallstring`](https://github.com/jFransham/smallstring) for that.

**Comment by OvermindDL1** [comment]
> @OvermindDL1 Regarding the cheap strings you were describing, we could use smallstring for that.

Different things.  That specific library you linked uses 1 byte per character, so you are limited to 8 characters if you want fast comparisons and lookups, where as a packed version you can get more characters in the 8 bytes.

Smallstring would not be bad though, just not the same thing as my Atoms.  ^.^

**Comment by OvermindDL1** [comment]
I think it would be more accurate to call one of my Atoms a kind of Global Enumeration, where every 'name' of the enumeration has a defined and unique integral value (either 32 bit or 64 bit depending on the context you are using) with no chance of collisions while being compile-time and efficient, with the restriction that it has a max length of 6 or 12 or so depending on the mapping and integer size.

**Comment by torkleyy** [comment]
@OvermindDL1 I see. Did you follow the (short) discussion on Gitter about the "event system"?

**Comment by OvermindDL1** [comment]
@torkleyy I had not, I can only have so many electron apps running with how heavy they are and Discord tends to keep up top since all of the other programming communities I'm in use it...  ^.^;

Is it summarized somewhere?

**Comment by torkleyy** [comment]
Yes, here, in the future :smile:

**Comment by Xaeroxe** [comment]
This appears to be a duplicate of #46 , we should close one of the two issues.

**Comment by Xaeroxe** [comment]
I'd really like to push this to completion as it could open some new doors for us in regards to input design and physics event handling.  So here's my two cents on this:

- Asynchronous event handling would require internal specs machinery that would probably end up being quite complicated to maintain, so I'd prefer to avoid that approach.
- Having the user post their own events could result in complications regarding execution order of states and systems.  I.E. things might execute a frame later than expected.

So I'd prefer for events to be exclusively a way for the engine to communicate to the end user.  To that end I have the following idea:

Add a new enum called Event

```Rust
pub enum Event {
    Physics(/*yada yada*/)
    Input(/*yada yada*/)
    /*Other event types*/
}
```

Add a new struct called Events

```Rust
pub struct Events {
    vec: Vec<Event>,
}
```

Then add this as a resource to the ECS world, and as needed internal systems can push events into the vector, then in the user systems and states they can call `.iter().filter()` on it to get the event types that they are interested in.  Once the frame is complete call `.clear()` then rinse and repeat.

Similarly we could also implement `Component` for `Events` and use that for entity specific events.

**Comment by omni-viral** [comment]
@Xaeroxe having `FetchMut<Events>` in multiple systems may decrease number of simultaneously running systems. Different types of events should be in separete vectors.

**Comment by Xaeroxe** [comment]
Well none of them should be `FetchMut`, my plan was they'd all be `Fetch`

**Comment by omni-viral** [comment]
@Xaeroxe no objections then )

**Comment by Xaeroxe** [comment]
With the integration of shrev to amethyst I consider this resolved.  Thanks @Rhuagh !
